### PR TITLE
Add first chapter solutions

### PR DESCRIPTION
**Pull request by julianespinel**
### Solutions for Chapter 1

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -477,8 +478,8 @@ Implement a function that returns the last digit of a given number.
   results. Or you can try to guess the function name, search for it and check
   whether it works for you!
 -}
--- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit n = (abs n) `mod` 10
```
Comment (chshersh) [inline src/Chapter1.hs]: Nice work on using infix form of calling functions! We don't explain this syntax in our course to make it short. But well done :+1: 

Also, you can remove `()`, they are redundant in this case :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Chapter1.hs [hunk 2]
```diff
@@ -542,7 +545,28 @@ value after "=" where the condition is true.
...
+sortSortedLists (x:xs) (y:ys)
+  | x <= y    = x : sortSortedLists xs (y:ys)
+  | otherwise = y : sortSortedLists (x:xs) ys
+
+
+mergeSort :: [Int] -> [Int]
+mergeSort [] = []
+mergeSort (x:[]) = [x]
+mergeSort (x:y:[]) = min x y : max x y : []
+mergeSort xs = do
```
Comment (chshersh) [inline src/Chapter1.hs:560]: Very nice work on implementing your own Merge Sort Algorithm!

File: src/Chapter1.hs [hunk 3]
```diff
@@ -542,7 +545,28 @@ value after "=" where the condition is true.
 Casual reminder about adding top-level type signatures for all functions :)
 -}
 
-mid x y z = error "mid: not implemented!"
+sortSortedLists :: [Int] -> [Int] -> [Int]
+sortSortedLists xs [] = xs
+sortSortedLists [] ys = ys
+sortSortedLists (x:xs) (y:ys)
```
Comment (chshersh) [inline src/Chapter1.hs]: I would rename this function to `mergeSorted` :slightly_smiling_face:

File: src/Chapter1.hs [hunk 4]
```diff
@@ -542,7 +545,28 @@ value after "=" where the condition is true.
...
+
+mergeSort :: [Int] -> [Int]
+mergeSort [] = []
+mergeSort (x:[]) = [x]
+mergeSort (x:y:[]) = min x y : max x y : []
+mergeSort xs = do
+  let sortedFirstHalf = mergeSort (fst halves)
+  let sortedSecondHalf = mergeSort (snd halves)
+  sortSortedLists sortedFirstHalf sortedSecondHalf
+    where halves = splitAt (length xs `div` 2) xs
```
Comment (chshersh) [inline src/Chapter1.hs:564]: That's a valid implementing of splitting the list into two halves. However, calculating the length of a list is a slow operation, as it traverses the whole list. In this particular case it is possible to split the list into two equal size lists in a single pass — elements on odd positions and elements on even positions.

File: src/Chapter1.hs [hunk 5]
```diff
@@ -542,7 +545,28 @@ value after "=" where the condition is true.
...
+
+
+mid :: Int -> Int -> Int -> Int
+mid x y z = sorted !! 1
+  where sorted = mergeSort [x, y, z]
```
Comment (chshersh) [inline src/Chapter1.hs:569]: Just one minor addition: you can implement this even simpler using `let in` and pattern-matching directly:

```haskell
mid x y z = let [_, m, _] = mergeSort [x, y, z] in m
```

File: src/Chapter1.hs [hunk 6]
```diff
@@ -556,7 +580,10 @@ True
 >>> isVowel 'x'
 False
 -}
-isVowel c = error "isVowel: not implemented!"
+
+isVowel :: Char -> Bool
+isVowel c = elem c vowels
```
Comment (chshersh) [inline src/Chapter1.hs:585]: Absolutely correct! :+1: 

One note in here, that sometimes, `elem` could be slower than the explicit pattern matching. I remember there were some benchmarks on one particular case, that showed how moving to pattern matching on each case separately drastically decrease time 🐎

File: src/Chapter1.hs [hunk 7]
```diff
@@ -620,7 +647,11 @@ Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
 
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Int -> Int
+sumLast2 n = currentDigit + nextDigit
+  where positive = abs n
+        currentDigit = positive `mod` 10
+        nextDigit    = (positive `div` 10) `mod` 10
```
Comment (chshersh) [inline src/Chapter1.hs:654]: On a side note, there is a standard function as well that combines both `div` and `mod` at once: `divMod`.

It returns the pair, so you can pattern match on that and get both variabless at once:

```haskell
let (a, b) = abs n `divMod` 10
```



That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: the standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that is actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(first, second) = divMod (mod x 100) 10
```

You can see how we could pattern match on the pair 🙂

**Review by chshersh — APPROVED** [review]
Awesome solutions to all tasks!
And well done on cracking the Advanced Task :trophy: 

I left some comments and feedback on solutions. But you nailed it :muscle: 

⚠️ By the way, per new rules of the Hacktoberfest event, you would need to add the `hacktoberfest` topic to this fork repo:

* https://hacktoberfest.digitalocean.com/hacktoberfest-update

All previous contributions are legit, but from now on it should have the label to participate.

**Comment by julianespinel** [comment]
@chshersh thank you so much for your feedback. I will go head with chapter 2!

**Pull request merged**